### PR TITLE
Add diagnostics reporting

### PR DESCRIPTION
**Pull request by Hoverbear**
##### Description

Add preliminary diagnostic reporting about pass/fail on uninstall/install attempts. This is a generic implementation that can report to any url or file path and contains minimal information.

##### Checklist

- [ ] Formatted with `cargo fmt`
- [ ] Built with `nix build`
- [ ] Ran flake checks with `nix flake check`
- [ ] Added or updated relevant tests (leave unchecked if not applicable)
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)
- [ ] Linked to related issues (leave unchecked if not applicable)

##### Validating with `install.determinate.systems`

If a maintainer has added the `upload to s3` label to this PR, it will become available for installation via `install.determinate.systems`:

```shell
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/pr/$PR_NUMBER | sh -s -- install
```

### CODE REVIEW

File: src/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+use os_release::OsRelease;
+use reqwest::Url;
+
+/// The static of an action attempt
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub enum DiagnosticStatus {
+    Cancelled,
+    Success,
+    /// This includes the [`strum::IntoStaticStr`] representation of the error, we take special care not to include parameters of the error (which may include secrets)
+    Failure(String),
```
Comment (cole-h) [inline src/diagnostics.rs:19]: I wonder, could we easily include which step failed as well...? Knowing it was a permissions error would be more useful if it also said "CreateFile".

**Review by cole-h — APPROVED** [review]
Aside from my comment (I don't consider it a blocker), this LGTM.

File: src/diagnostics.rs (hunk 1 above)
Comment (Hoverbear) [reply src/diagnostics.rs:19]: Yeah... It's something I'd like to improve. I think we have some work to do on error reporting in general to be honest. I'd like to wrap that into the later work.

Comment (cole-h) [reply src/diagnostics.rs:19]: Sounds good to me. I opened https://github.com/DeterminateSystems/nix-installer/issues/268 for that.

**Pull request merged**